### PR TITLE
KAFKA-16535: Implement AddVoter, RemoveVoter, UpdateVoter RPCs

### PR DESCRIPTION
**Pull request by cmccabe**
Implement the add voter, remove voter, and update voter RPCs for
KIP-853. This is just adding the RPC handling; the current
implementation in RaftManager just throws UnsupportedVersionException

### CODE REVIEW

**Comment by AndrewJSchofield** [comment]
LGTM apart from a few test failures related to the new RPCs, such as RequestConvertToJsonTest.

File: clients/src/main/resources/common/message/AddRaftVoterRequest.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "name": "AddRaftVoterRequest",
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "ClusterId", "type": "string", "versions": "0+" },
+    { "name": "TimeoutMs", "type": "int32", "versions": "0+" },
+    { "name": "VoterId", "type": "int32", "versions": "0+",
+      "about": "The replica id of the voter getting added to the topic partition" },
+    { "name": "VoterDirectoryId", "type": "uuid", "versions": "0+",
+      "about": "The directory id of the voter getting added to the topic partition" },
```
Comment (jsancio) [inline clients/src/main/resources/common/message/AddRaftVoterRequest.json:29]: @cmccabe and I discussed this offline and agreed to remove topic name, topic id and topic partition from the requests. Kafka only supports one KRaft partition (__cluster_metadata-0). If we decide to have more than one KRaft partition we can revisit this decision with another KIP.

I will update KIP-853 to match these changes once this PR is merged.

File: clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java [hunk 2]
```diff
@@ -1320,6 +1320,16 @@ public ListClientMetricsResourcesResult listClientMetricsResources(ListClientMet
...
 
+    @Override
+    public AddRaftVoterResult addRaftVoter(int voterId, Uuid voterDirectoryId, Set<RaftVoterEndpoint> endpoints, AddRaftVoterOptions options) {
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
+
+    @Override
+    public RemoveRaftVoterResult removeRaftVoter(int voterId, Uuid voterDirectoryId, RemoveRaftVoterOptions options) {
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
```
Comment (jsancio) [inline clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:1331]: I suspect that we will never implement these. Do you agree?

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 3]
```diff
@@ -318,4 +329,28 @@ class KafkaRaftManager[T](
   override def voterNode(id: Int, listener: String): Option[Node] = {
     client.voterNode(id, listener).toScala
   }
+
+  def verifyCommonFields(
+    providedClusterId: String,
+  ): Unit = {
+    if (!providedClusterId.equals(clusterId)) {
+      throw new InconsistentClusterIdException("Provided cluster ID " + providedClusterId +
```
Comment (jsancio) [inline core/src/main/scala/kafka/raft/RaftManager.scala]: This applies to all of the `throw new` added.

If we throw here, this will get caught here and print an error message, right?
https://github.com/apache/kafka/pull/16058/files#diff-91060c918c99d25342f625c146f14425716eda9d8fcfe1126b2c45feff388362R152-R154

How about for now we just complete the requests with the unsupported version error responses?

Also, I have to double check but I think cluster ids are check by the `KafkaRaftClient` and not the `KafkaRaftManager`. We have this abstraction so we can test that behavior in the `KafkaRaftClientTest` suite or if we add a new test suite for this functionality.

Most our interesting tests are against `KafkaRaftClient`. The tests for `KafkaRaftManager` are more basic and limited.

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 4]
```diff
@@ -1080,4 +1083,39 @@ class ControllerApis(
...
+  def handleAddRaftVoter(request: RequestChannel.Request): CompletableFuture[Unit] = {
+    authHelper.authorizeClusterOperation(request, ALTER)
+    val addRequest = request.body[AddRaftVoterRequest]
+    raftManager.handleAddVoter(addRequest.data())
+    requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs =>
+      new AddRaftVoterResponse(new AddRaftVoterResponseData().
+        setThrottleTimeMs(requestThrottleMs).
+        setErrorCode(0.toShort).
+        setErrorMessage(null)))
+    CompletableFuture.completedFuture[Unit](())
```
Comment (jsancio) [inline core/src/main/scala/kafka/server/ControllerApis.scala]: I think this is going to change in the real implementation. For KRaft requests we simply forward them to the raft thread/actor using `handleRaftRequest` and `raftManager.handleRequest`. 

How about for now we just send an error response with "unsupported version" and return a completed future?

This comment applies to all of the new methods added to this type.

File: core/src/test/resources/log4j.properties [hunk 5]
```diff
@@ -18,8 +18,8 @@ log4j.appender.stdout=org.apache.log4j.ConsoleAppender
 log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
 
-log4j.logger.kafka=WARN
-log4j.logger.org.apache.kafka=WARN
+log4j.logger.kafka=OFF
+log4j.logger.org.apache.kafka=OFF
```
Comment (jsancio) [inline core/src/test/resources/log4j.properties]: Why this change? Did you mean to change this?

File: core/src/test/scala/unit/kafka/raft/RaftManagerTest.scala [hunk 6]
```diff
@@ -306,6 +309,113 @@ class RaftManagerTest {
 
   }
 
+  class ReconfigurationTestContext(
```
Comment (jsancio) [inline core/src/test/scala/unit/kafka/raft/RaftManagerTest.scala]: If you agree with my comments in `RaftManager`, I am okay removing these tests. I think when we have the real implementation will want to add a `KafkaRaftClientReconfigTest`, similar to `KafkaRaftClientTest` and `KafkaRaftClientSnapshotTest` that uses `RaftClientTestContext` to test `KafkaRaftClient`.

File: clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * An endpoint for a raft quorum voter.
+ */
+@InterfaceStability.Stable
+public class RaftVoterEndpoint {
+    private final String name;
+    private final String host;
+    private final int port;
+    private final String securityProtocol;
```
Comment (jsancio) [inline clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java]: Did you consider using `o.a.k.c.Endpoints`? This struct has a similar structure to `Endpoints`.

During the KIP discussion we agreed to not add security protocol to the `AddRaftVoter` request. Mainly because it is not needed and that there is a requirement that the listener name, security protocol and the protocol configuration must match in all of the voters if they exist.

**Review by jsancio — COMMENTED** [review]
Thanks for the changes @cmccabe

File: clients/src/main/resources/common/message/AddRaftVoterRequest.json (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/resources/common/message/AddRaftVoterRequest.json:29]: Right. And it would be easy to update the RPCs to have an optional topic structure, if and when we need one.

File: clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java (hunk 2 above)
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:1331]: Maybe not. But Java requires them to be in MockAdminClient so that it will typecheck.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 3 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/raft/RaftManager.scala]: > If we throw here, this will get caught here and print an error message, right?
https://github.com/apache/kafka/pull/16058/files#diff-91060c918c99d25342f625c146f14425716eda9d8fcfe1126b2c45feff388362R152-R154

Hmm, your URL doesn't work for me. As a general matter, exceptions not otherwise handled in ControllerApis.scala will just invoke `AbstractRequest.getErrorResponse`.

> How about for now we just complete the requests with the unsupported version error responses?

OK

File: core/src/test/resources/log4j.properties (hunk 5 above)
Comment (cmccabe) [reply core/src/test/resources/log4j.properties]: Sorry, will fix

File: clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java (hunk 7 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java]: > Did you consider using o.a.k.c.Endpoints? This struct has a similar structure to Endpoints.

`org.apache.kafka.common.Endpoint` is not intended to be used as part of the Admin Client API. That class drags in the non-public class `org.apache.kafka.common.security.auth.SecurityProtocol`. Additionally we may want to change it in the future, which we could not do if it were part of the admin API.

(`org.apache.kafka.common.Endpoint` did get dragged into the Authorizer API, which I think was a mistake, but that's a separate discussion. And not many people have ever written Authorizers, so changing that stuff is not impossible)

> During the KIP discussion we agreed to not add security protocol to the AddRaftVoter request. Mainly because it is not needed and that there is a requirement that the listener name, security protocol and the protocol configuration must match in all of the voters if they exist.

Is the expectation that the security protocol is found by looking up the endpoint name in `listener.security.protocol.map`? What happens if it's not there? Do we fail or fall back to `PLAINTEXT`?

File: core/src/main/scala/kafka/server/ControllerApis.scala (hunk 4 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/ControllerApis.scala]: ok

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 3 above)
Comment (jsancio) [reply core/src/main/scala/kafka/raft/RaftManager.scala]: > Hmm, your URL doesn't work for me.

This is the URL/code I was trying to send: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ControllerApis.scala#L147-L151

File: clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java (hunk 7 above)
Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/clients/admin/RaftVoterEndpoint.java]: > Is the expectation that the security protocol is found by looking up the endpoint name in `listener.security.protocol.map`? What happens if it's not there? Do we fail or fall back to `PLAINTEXT`?

It will use whatever is returned by `KafkaConfig.effectiveListenerSecurityProtocolMap`. Yes, it does look like that code currently returns PLAINTEXT if there is not security mapping for the controller listener.

File: core/src/test/scala/unit/kafka/raft/RaftManagerTest.scala (hunk 6 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/raft/RaftManagerTest.scala]: ok

**Review by jsancio — APPROVED** [review]
LGTM. The failures look unrelated to me. What do you think @cmccabe

**Comment by cmccabe** [comment]
thanks, committed.
